cmd/sriovdp/manager.go

```diff
@@ -143,21 +159,122 @@ func (rm *resourceManager) validConfigs() bool {
-			for _, addr := range pciAddrs {
-				if configured := utils.SriovConfigured(addr); !configured {
-					glog.Errorf("no SRIOV VF configured for root device %s", addr)
-					return false
+		resourceName[conf.ResourceName] = conf.ResourceName
+	}
+
+	return true
+}
+
+func (rm *resourceManager) discoverHostDevices() error {
+	glog.Infoln("discovering host network devices")
+
+	pci, err := ghw.PCI()
+	if err != nil {
+		return fmt.Errorf("discoverDevices(): error getting PCI info: %v", err)
+	}
+
+	devices := pci.ListDevices()
+	if len(devices) == 0 {
+		return fmt.Errorf("discoverDevices(): error: could not retrive PCI info: %v,", err)
+	}
+	for _, device := range devices {
+		devClass, err := strconv.ParseInt(device.Class.ID, 16, 64)
+		if err != nil {
+			return fmt.Errorf("discoverDevices(): error parsing device class %v", err)
+		}
+
+		// only interested in network class
+		if devClass == netClass {
+			vendor := device.Vendor
+			vendorName := vendor.Name
+			if len(vendor.Name) > 20 {
+				vendorName = string([]byte(vendorName)[0:17]) + "..."
+			}
+			product := device.Product
+			productName := product.Name
+			if len(product.Name) > 40 {
+				productName = string([]byte(productName)[0:37]) + "..."
+			}
+			glog.Infof("discoverDevices(): device found: %-12s\t%-12s\t%-20s\t%-40s", device.Address, device.Class.ID, vendorName, productName)
+
+			// exclude device in-use in host
+			if isUsed, _ := isInUse(device.Address); !isUsed {
+
+				aPF := utils.IsSriovPF(device.Address)
+				aVF := utils.IsSriovVF(device.Address)
+
+				if aPF || !aVF {
+					// add to linkWatchList
+					rm.addToLinkWatchList(device.Address)
+				}
+
+				if aPF && utils.SriovConfigured(device.Address) {
+					// do not add this device in net device list
+					continue
+				}
+
+				if newDevice, err := resources.NewPciNetDevice(device, rm.rFactory); err == nil {
+					rm.netDeviceList = append(rm.netDeviceList, newDevice)
+				} else {
+					glog.Errorf("discoverDevices() error adding new device: %q", err)
 				}
+
 			}
 		}
+	}
+	return nil
+}
 
-		// [To-Do]: Validate deviceType
+// isInUse returns true if PCI network device appear to be in use by host system given its pci address as string.
+// Otherwise it is assumed not to be in used.
+func isInUse(pciAddr string) (bool, error) {
 
-		resourceName[conf.ResourceName] = conf.ResourceName
+	// inUse := false
+	// Get net interface name
+	ifNames, err := utils.GetNetNames(pciAddr)
+	if err != nil {
+		return false, fmt.Errorf("error trying get net device name for device %s", pciAddr)
 	}
 
-	return true
+	if len(ifNames) > 0 { // there's at least one interface name found
+		for _, ifName := range ifNames {
+			link, err := netlink.LinkByName(ifName)
+			if err != nil {
+				glog.Errorf("expected to get valid host interface with name %s: %q", ifName, err)
+			}
+
+			routes, err := netlink.RouteList(link, netlink.FAMILY_V4) // IPv6 routes: all interface has at least one link local route entry
```

Comment:
Is there way to detect if vfio/uio devices are in use?

Reply:
I couldn't think of any other than querying the kubelet resource api. This may get little bit complicated